### PR TITLE
Serverside search and postloading fact_names to reduce client load

### PR DESCRIPTION
**Pull request by Marius2805**
In large environments with huge amount of facts/fact_names the performance when rendering the /facts page and client side search is very bad.
For example in our enterprise environment we have > 90k facts defined and it takes up to 2 mins to load/render the /facts page. Searching is not possible at all (100% CPU Load for several minutes).

Therefore this branch contains:
- postloading feature for /facts page, that means facts are loaded in 1k steps. If the user enters the end of the page the next 1k fact_names will be postloaded via AJAX.
- server side search feature. Search operations are performed on server Results will be loaded via AJAX. Server side search is much more faster than client side JQuery search (for 90k < 100ms) 

To enable proper postloading while scrolling, layout of /facts page has been changed from column view to vertical table view.

### CODE REVIEW

File: puppetboard/app.py [hunk 1]
```diff
@@ -260,11 +260,17 @@ def report(node_name, report_id):
...
+@app.route('/facts/<start>')
+@app.route('/facts/<start>/<end>')
+def facts(start=0, end=1000):
     """Displays an alphabetical list of all facts currently known to
-    PuppetDB."""
+    PuppetDB. To not overload the client rendering engine, output is limited to 1000. """
     facts_dict = collections.defaultdict(list)
     facts = get_or_abort(puppetdb.fact_names)
+    start = int(float(start))
+    end   = int(float(end))
```
Comment (daenney) [inline puppetboard/app.py]: Why the `int(float))` casting?

Comment (Marius2805) [inline puppetboard/app.py]: Thank you. Was just a relic of testing an different approach. Replaced with normal int() casting.

**Comment by Marius2805** [comment]
Maybe you didn´t noticed my commit to fix the unnecessary casting, therefore I want to ask for a feedback of the changes.

**Comment by fgimian** [comment]
Hey there @Marius2805, I haven't yet been able to spend some time going through all pull requests here but for bigger ones like this, I'll likely ask for @daenney to review it with me after I have reviewed it.

For now, can you please repair your indentation in the code?  It seems you are using tabs.  We preofer soft spaces (no tabs) and that you kindly follow the indentation style used in the files already.

Thanks a lot for your contribution
Fotis

**Comment by Marius2805** [comment]
Hello @fgimian thank you for your investigation.
In the meantime I replaced all tabs with soft spaces to follow your coding style.

**Comment by jyaworski** [comment]
Hello:

Is this still needed? If so, can you rebase and squash?

**Comment by Marius2805** [comment]
@sharlander You have many facts and servers in your environment, could you please test if the current master branch solved already the performance issues?
If not I will rebase this branch.

**Comment by corey-hammerton** [comment]
Closing due to inactivity.

**Pull request closed without merging**